doom/config.org
#+title: DOOM Emacs Config
#+STARTUP: showeverything
#+PROPERTY: header-args:emacs-lisp :tangle yes :cache yes :results silent
#+export_file_name: emacs
#+hugo_weight: 2

* Table of Contents :toc:noexport:
- [[#about-this-config][About this Config]]
- [[#basics][Basics]]
  - [[#sensible-defaults][Sensible Defaults]]
  - [[#emacs-everywhere][Emacs EVERYWHERE]]
- [[#appearance][Appearance]]
  - [[#fonts][Fonts]]
  - [[#theme][Theme]]
  - [[#modeline][Modeline]]
  - [[#dashboard-startup-page][Dashboard (startup page)]]
  - [[#line-numbers-wrapping-and-margins][Line Numbers, Wrapping and Margins]]
  - [[#zenwriteroom][Zen/Writeroom]]
  - [[#transparent-background][Transparent Background]]
  - [[#whitespace][Whitespace]]
  - [[#treemacs][Treemacs]]
- [[#global-functionality][Global Functionality]]
  - [[#local-leader][Local Leader]]
  - [[#movecut][Move/Cut]]
  - [[#windows-and-buffers][Windows and Buffers]]
  - [[#auto-complete][Auto-complete]]
  - [[#search--replace][Search & Replace]]
  - [[#spell--and-grammar-checker][Spell- and grammar checker]]
  - [[#snippets][Snippets]]
  - [[#performance][Performance]]
  - [[#debugging][Debugging]]
- [[#org-mode][Org Mode]]
  - [[#org-paths][Org: Paths]]
  - [[#org-appearance][Org: Appearance]]
  - [[#org-organizational-settings][Org: Organizational Settings]]
  - [[#org-roam][Org Roam]]
  - [[#org-latex-previews][Org LaTeX previews]]
  - [[#xournal-integration-org-notebook][Xournal++ integration ("Org Notebook")]]
  - [[#org-d20][org-d20]]
  - [[#ox-hugo][ox-hugo]]
- [[#latex][LaTeX]]
  - [[#fixing-defaults]["Fixing" defaults]]
  - [[#viewer][Viewer]]
  - [[#appearance-1][Appearance]]
  - [[#cdlatex][CDLaTeX]]
  - [[#reftex][RefTeX]]
- [[#literature--citations][Literature & Citations]]
  - [[#note-taking][Note Taking]]
  - [[#appearance-2][Appearance]]
  - [[#keybindings][Keybindings]]
  - [[#pdf-tools][PDF-Tools]]
  - [[#org-noter][Org Noter]]
- [[#tangle-this-file][Tangle this file!]]

* About this Config
Here is some help from the default config.el that I haven't found a better spot for yet.
#+begin_quote
Here are some additional functions/macros that could help you configure Doom:

- `load!' for loading external *.el files relative to this one
- `use-package!' for configuring packages
- `after!' for running code after a package has loaded
- `add-load-path!' for adding directories to the `load-path', relative to
  this file. Emacs searches the `load-path' when you load packages with
  `require' or `use-package'.
- `map!' for binding new keys

To get information about any of these functions/macros, move the cursor over
the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
This will open documentation for it, including demos of how they are used.

You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
they are implemented.
#+end_quote


* Basics

#+begin_src emacs-lisp
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

(setq user-full-name "Michael Reitmeir"
      user-mail-address "michi.reitmeir@gmail.com")
#+end_src

** Sensible Defaults
Inspired by [[https://tecosaur.github.io/emacs-config/config.html#better-defaults][tecosaur]], who was in turn inspired by others. Funny how that works.
#+begin_src emacs-lisp
(setq-default
 delete-by-moving-to-trash t                      ; Delete files to trash
 window-combination-resize t                      ; take new window space from all other windows (not just current)
 x-stretch-cursor t)                              ; Stretch cursor to the glyph width

(setq undo-limit 80000000                         ; Raise undo-limit to 80Mb
      evil-want-fine-undo t                       ; By default while in insert all changes are one big blob. Be more granular
      auto-save-default t                         ; Nobody likes to loose work, I certainly don't
      truncate-string-ellipsis "…"
      completion-ignore-case t)

(setq evil-vsplit-window-right t
      evil-split-window-below t)

(defconst doom-module-config-file (concat doom-user-dir "config.org"))
(defun open-emacs-config-file () "open `doom-module-config-file'"
  (interactive)
  (find-file doom-module-config-file))
(map! :leader
 (:prefix ("o" . "open")
       :desc "Emacs config"     "c"     #'open-emacs-config-file))
#+end_src

The [[https://fishshell.com/][fish]] shell is not [[https://stackoverflow.com/questions/48732986/why-how-fish-does-not-support-posix][POSIX compatible]], which among other things causes a bunch of garbage characters to appear whenever it is used in emacs (at least with my fish config). So it's better to have emacs use bash instead, especially in the interest of packages that rely on shell outputs
#+begin_src emacs-lisp :tangle yes
(setq shell-file-name (executable-find "bash"))
#+end_src
** Emacs EVERYWHERE
This changes the name of windows created with ~emacs-everwhere~, so that I can more easily customize their behaviour in my window manager.
#+begin_src emacs-lisp :tangle yes
(setq emacs-everywhere-frame-name-format "emacs-everywhere")
#+end_src
Specifically, I want my window manager to handle it's position rather than emacs.
#+begin_src emacs-lisp :tangle yes
(remove-hook 'emacs-everywhere-init-hooks #'emacs-everywhere-set-frame-position)
#+end_src
Org mode is nice. But the variable =emacs-everywhere-major-mode-function= itself doesn't seem to do the trick. So let's add org to the init hook as well.
#+begin_src emacs-lisp :tangle yes
(setq emacs-everywhere-major-mode-function #'org-mode)
(add-hook 'emacs-everywhere-init-hooks 'org-mode)
#+end_src

* Appearance
** Fonts
From the default config.el:
#+begin_quote
Doom exposes five (optional) variables for controlling fonts in Doom. Here
are the three important ones:

+ `doom-font'
+ `doom-variable-pitch-font'
+ `doom-big-font' -- used for `doom-big-font-mode'; use this for
  presentations or streaming.

They all accept either a font-spec, font string ("Input Mono-12"), or xlfd
font string. You generally only need these two:
(setq doom-font (font-spec :family "monospace" :size 12 :weight 'semi-light)
      doom-variable-pitch-font (font-spec :family "sans" :size 13))
#+end_quote

#+begin_src emacs-lisp
(setq doom-font (font-spec :family "JetBrains Mono" :size 15 :weight 'regular)
      doom-variable-pitch-font (font-spec :family "Libertinus Sans" :size 19))
;;      doom-variable-pitch-font (font-spec :family "Fira Sans") ; inherits `doom-font''s :size
;;      doom-unicode-font (font-spec :family "Input Mono Narrow" :size 12)
;;      doom-big-font (font-spec :family "Fira Mono" :size 19))

(setq-default line-spacing 0.25)
#+end_src

I also gotta activate ~mixed-pitch-set-height~, so that ~mixed-pitch-mode~ actually uses the ~:size~ option from the variable pitch font above.
#+begin_src emacs-lisp :tangle yes
(setq mixed-pitch-set-height t)
#+end_src

Let's bind some keys for quick toggling between the different font modes.
#+begin_src emacs-lisp
(map! :leader
 (:prefix ("t" . "toggle")
       :desc "Mixed pitch mode"       "m"     #'mixed-pitch-mode
       :desc "Variable pitch mode"    "v"     #'variable-pitch-mode
       )
      )
#+end_src
By this, ~visible-mode~ has been unmapped from ~SPC t v~. Let's remap it to ~SPC t V~.
#+begin_src emacs-lisp
(map! :leader
 (:prefix ("t" . "toggle")
       :desc "Visible mode"           "V"     #'visible-mode
       )
      )
#+end_src

** Theme
I like DOOMs default colors
#+begin_src emacs-lisp
(setq doom-theme 'doom-one)
#+end_src

But I still wanna customize it a bit.
*** Shades with ewal
I use [[https://github.com/dylanaraps/pywal][pywal]] to have the colors in my terminal fit my wallpaper.
While for Emacs I use the colors from the ~doom-one~ theme, I still want the (transparent) backgrounds of different parts of Emacs to synergize with pywal.
For this, the [[https://github.com/cyruseuros/ewal][ewal]] package is used to get the color scheme from pywal into Emacs.

The faces and their desired changes are stored in ~ewal-background-list~. The function ~toggle-ewal-background~ then loops over this list and applies the changes, unless ~ewal-background-active~ is non-nil, in which case it reloads the theme to undo the changes.

#+begin_src emacs-lisp
(use-package ewal)
(use-package ewal-doom-themes)

(defvar ewal-background-list
  '((default :background (ewal-load-color 'background))
    (separator-line :background (ewal-load-color 'background))
    (hl-line :background (ewal--color-chshade (ewal-load-color 'background) .1))
    (org-block :background (ewal--color-chshade (ewal-load-color 'background) -0.3))
        ;; Tabs:
    (tab-bar :background (ewal-load-color 'background))
    (tab-bar-tab :background (ewal--color-chshade (ewal-load-color 'background) .1))
    (tab-bar-tab-inactive :background (ewal--color-chshade (ewal-load-color 'background) .05))
    (tab-line :background (ewal-load-color 'background))
        ;; Mode line:
    (mode-line :background (ewal--color-chshade (ewal-load-color 'background) .15))
    (mode-line-inactive :background (ewal--color-chshade (ewal-load-color 'background) .05))
    (mode-line-emphasis :background (ewal--color-chshade (ewal-load-color 'background) .20))
        ;; minibuffer (underneath mode line) and stuff
    (solaire-default-face :background (ewal-load-color 'background)))
  "list of faces to customize when styling emacs with ewal")

(defvar ewal-background-active nil "non-nil if background is currently styled using ewal")

(defun toggle-ewal-background ()
  "toggle ewal background colors on and off"
  (interactive)
  (if ewal-background-active
        (doom/reload-theme)
        (dolist (spec ewal-background-list)
                (let ((face (nth 0 spec))
                (attribute (nth 1 spec))
                (value (nth 2 spec)))
                (set-face-attribute face nil attribute (eval value)))))
  (setq ewal-background-active (not ewal-background-active)))

(map! :leader
 (:prefix ("t" . "toggle")
       :desc "background colors"        "B"     #'toggle-ewal-background))
#+end_src

Let's turn on the ewal background by default.
#+begin_src emacs-lisp
(after! (doom-themes org hl-line)
        (toggle-ewal-background))
#+end_src

** Modeline
Making the modeline a bit taller than the default height (25) seems nice.
#+begin_src emacs-lisp
(setq doom-modeline-height 35)
#+end_src

** Dashboard (startup page)
I like this silly banner I found at [[https://github.com/jeetelongname/doom-banners]] for my dashboard. To fit with this, font color is changed to pink.
#+begin_src emacs-lisp
(after! doom-themes
    (custom-theme-set-faces! 'doom-one
        `(doom-dashboard-banner :foreground "pink" :weight bold)))
(setq fancy-splash-image "~/.config/doom/I-am-doom.png")
(setq +doom-dashboard-banner-padding '(0 . 4))
#+end_src

Now let's change the menu options on the dashboard. I added ~doom-dashboard-widget-projects~, which contains extra buttons for all projects known to projectile. The source code is adapted from the ~doom-dashboard-widget-shortmenu~ function.
#+begin_src emacs-lisp
(setq +doom-dashboard-functions
      '(doom-dashboard-widget-banner
        doom-dashboard-widget-shortmenu
        doom-dashboard-widget-projects))

(setq +doom-dashboard-menu-sections
 '(("Configure Emacs"
    :icon (concat (nerd-icons-icon-for-mode 'emacs-lisp-mode :face 'doom-dashboard-menu-title) " ")
    :action open-emacs-config-file
    :key "SPC o c")
   ("Take some notes"
    :icon (concat (nerd-icons-faicon "nf-fa-file_pen" :face 'doom-dashboard-menu-title) " ")
    :action org-roam-node-find-default
    :key "SPC r f")
   ("Read some literature"
    :icon (concat (nerd-icons-faicon "nf-fa-book" :face 'doom-dashboard-menu-title) " ")
    :action citar-open-files
    :key "SPC l f")
   ("Work on a project"
    :icon (concat (nerd-icons-faicon "nf-fa-code" :face 'doom-dashboard-menu-title) " ")
    :action projectile-switch-project
    :key "SPC p p")))

(require 'projectile)
(defface doom-dashboard-project
  '((t :inherit font-lock-string-face))
  "Face for projects on doom-dashboard.")
(defun doom-dashboard-widget-projects ()
  "Create list of menu buttons for each project in `projectile-known-projects', and bind key appropriately."
        (projectile-known-projects)
        (dolist (project projectile-known-projects)
         (insert
          (+doom-dashboard--center (- +doom-dashboard--width 1)
           ; put together three things:
           (format "%s %-35s%-2s\n"
            ; 1. the icons
            (nerd-icons-faicon "nf-fa-angle_right" :face 'doom-dashboard-project)
            ; 2. the project button
            (with-temp-buffer
             (insert-text-button project
                        'action `(lambda (_button) (let
                                ((projectile-switch-project-action 'magit-status))
                                (projectile-switch-project-by-name ,project)))
                        'face 'doom-dashboard-project
                        'follow-link t
                        'help-echo project)
             (buffer-string))
            ; 3. the number (with face and keybindings)
            (let* ((num (+ 1 (cl-position project projectile-known-projects)))
                  (str (number-to-string num)))
             (map! :map +doom-dashboard-mode-map :ng str
                   `(lambda () (interactive) (let
                                ((projectile-switch-project-action 'magit-status))
                                (projectile-switch-project-by-name ,project))))
             (add-text-properties 0 (length str) (list 'face 'doom-dashboard-menu-desc) str)
              str))))))
(after! (projectile doom)
    (setq! +workspaces-switch-project-function #'doom-project-browse)
    (add-to-list 'projectile-ignored-projects "~/.config/emacs/"))
#+end_src

The following is a hack to remove the blank lines between menu buttons. By investigating the last lines of the source code of ~doom-dashboard-widget-shortmenu~, one can see that the newlines are inserted if ~display-graphic-p~ returns true. So let's temporarily overwrite this function with ~ignore~, which always returns nil.
#+begin_src emacs-lisp
(defadvice! no-new-lines (oldfun)
  :around #'doom-dashboard-widget-shortmenu
  (cl-letf (((symbol-function 'display-graphic-p) #'ignore))
    (funcall oldfun)))
#+end_src

Let's also add some shorter keybindings to the dashboard.
#+begin_src emacs-lisp
(map! :map +doom-dashboard-mode-map
      :ng "c"       #'open-emacs-config-file
      :ng "r"       #'org-roam-node-find-default
      :ng "l"       #'citar-open-files
      :ng "p"       #'projectile-switch-project)
#+end_src

** Line Numbers, Wrapping and Margins
Display relative line numbers, but do so counting lines as displayed, not actual line breaks in the buffer.
#+begin_src emacs-lisp
(setq display-line-numbers-type 'visual)
#+end_src
This works well for me, because I like overlength lines to always automatically wrap.
#+begin_src emacs-lisp
(global-visual-line-mode t)
#+end_src

qtile/config.py
# Copyright (c) 2010 Aldo Cortesi
# Copyright (c) 2010, 2014 dequis
# Copyright (c) 2012 Randall Ma
# Copyright (c) 2012-2014 Tycho Andersen
# Copyright (c) 2012 Craig Barnes
# Copyright (c) 2013 horsik
# Copyright (c) 2013 Tao Sauvage
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

from libqtile import bar, hook, qtile, layout, widget
from libqtile.config import Click, Drag, DropDown, Group, Key, Match, ScratchPad, Screen
from libqtile.lazy import lazy
import json
import os
import subprocess
import colorsys

hostname = subprocess.Popen("hostname", stdout=subprocess.PIPE ).communicate()[0]
hostname = hostname.strip()  # remove trailing newline
hostname = hostname.decode("utf-8")  # decode from type 'byte' to type 'str'

if hostname == "One":
    config_in_use = "desktop"
    desktop = True
    laptop = False
else:
    config_in_use = "laptop"
    desktop = False
    laptop = True

process = subprocess.Popen('~/.config/change-wallpaper.sh', shell=True, stdout=subprocess.PIPE)
process.wait()

colors = os.path.expanduser('~/.cache/wal/colors.json')
colordict = json.load(open(colors))
wal_foreground = colordict['special']['foreground']
wal_background = colordict['special']['background']
wal_cursor = colordict['special']['cursor']
wal_colors = [colordict['colors']['color' + str(i)] for i in range(16)]

def string_to_rgb(col):
    col = col.strip('#')
    if len(col) not in {6,8}:
        raise ValueError("Not a valid color string.")
    red = int(col[0:2], 16)/255
    green = int(col[2:4], 16)/255
    blue = int(col[4:6], 16)/255
    if len(col) == 8:
        alpha = int(col[6:8], 16)/255
        return (red,green,blue,alpha)
    else:
        return (red,green,blue)

def rgb_to_string(red,green,blue, alpha=None):
    col = [red,green,blue,alpha]
    if alpha == None:
        col.remove(None)
    if not all([0<=num<=1 for num in col]):
        raise ValueError("Color values must be between 0 and 1.")
    for i in range(len(col)):
        col[i] = hex(round(255*col[i]))[2:]
        while len(col[i]) < 2:
            col[i] = "0" + col[i]
    return ''.join(col)

def modify_helper(col, value, i):
    if type(col) == str:
        col_rgb = list(string_to_rgb(col))
    if type(col) == list:
        col_rgb = col.copy
    if type(col) == tuple:
        col_rgb = list(col)

    if len(col_rgb) == 4:
        alpha = col_rgb[3]
    else:
        alpha = None

    col_hsl = colorsys.rgb_to_hls(*col_rgb[0:3]) # remove alpha
    current_value = col_hsl[i]
    if type(value) == str:
        value = value.strip('%')
        perc = int(value)/100
        new_value = max(0, min(1, current_value*(1+perc)))
    else:
        new_value = float(value)
    col_hsl = list(col_hsl)
    col_hsl[i] = new_value
    col_rgb = colorsys.hls_to_rgb(*col_hsl)
    col_rgb = list(col_rgb)
    if alpha != None:
        col_rgb.append(alpha)

    if type(col) == str:
        return rgb_to_string(*col_rgb)
    if type(col) == list:
        return list(col_rgb)
    if type(col) == tuple:
        return tuple(col_rgb)


def modify_lightness(col,lightness):
    return modify_helper(col,lightness,1)
def modify_saturation(col,saturation):
    return modify_helper(col,saturation,2)

wal_colors_without_background = [col for col in wal_colors if col != wal_background]
wal_hls_without_background = [colorsys.rgb_to_hls(*string_to_rgb(col))
                              for col in wal_colors_without_background]

saturations = [colorsys.rgb_to_hls(*string_to_rgb(col))[2] for col in wal_colors_without_background]
wal_sorted_saturation = [pair[1] for pair in
                         sorted(zip(saturations, wal_colors_without_background),
                                key=lambda pair: -pair[0])]

lightnesses = [colorsys.rgb_to_hls(*string_to_rgb(col))[1] for col in wal_colors_without_background]
wal_sorted_lightness = [pair[1] for pair in
                         sorted(zip(lightnesses, wal_colors_without_background),
                                key=lambda pair: -pair[0])]

wal_background_hls = colorsys.rgb_to_hls(*string_to_rgb(wal_background))
hue_differences = [abs(col[0] - wal_background_hls[0]) for col in wal_hls_without_background]
wal_sorted_huediff = [pair[1] for pair in
                         sorted(zip(hue_differences, wal_colors_without_background),
                                key=lambda pair: -pair[0])]

wal_background_lightness = wal_background_hls[1]

# some magic numbers one can tweak
min_lightness = 0.5 * wal_background_lightness
max_lightness = min(2*wal_background_lightness,1)
num_lightness = 5

lightness_variations = [min_lightness + i*(max_lightness-min_lightness)/(num_lightness-1)
               for i in range(num_lightness)]
wal_background_versions = [modify_lightness(wal_background,l) for l in lightness_variations]

color_primary = modify_lightness(wal_sorted_saturation[0],0.7)
color_primary = modify_saturation(color_primary,"+20%")

color_secondary = modify_lightness(wal_sorted_saturation[1],0.85)
color_secondary = modify_saturation(color_secondary,"+20%")

bar_opacity = "C0"           # two hex digits
color_bar_background = wal_background + bar_opacity
color_bar_foreground = wal_colors[-1] + "FF"

follow_mouse_focus = True
bring_front_click = "floating_only"
floats_kept_above = True
cursor_warp = False
auto_fullscreen = True
focus_on_window_activation = "smart"
reconfigure_screens = True
auto_minimize = True
wl_input_rules = None
wmname = "qtile"

terminal = "kitty"
file_manager = "nemo"
web_browser = "firefox"
emacs = "emacsclient -c -a ''"
sysmon = terminal + " htop"
lockscreen = "xlock -mode \"rain\" -saturation 1 -mousemotion -timeout 10 -password \"Password please.\""
hibernate = ["sh", "-c", lockscreen + "& systemctl hibernate"]
suspend = ["sh", "-c", lockscreen + "& systemctl suspend"]

keys = []

layouts = [
    layout.MonadTall(
        border_width = 2,
        margin = 8,
        border_focus = color_primary,
        name = "",
        ),
    # layout.MonadWide(),
    layout.Max(name = ""),
]

floating_layout_theme = {"border_width": 2,
                "border_focus": color_secondary,
                "border_normal": wal_background}

groups = [Group(i) for i in "1234567890"]

def go_to_group(qtile,group_name):
    for s in qtile.screens:
        if s.group.name == group_name:
            qtile.cmd_to_screen(qtile.screens.index(s))
            return
    qtile.groups_map[group_name].toscreen()

def switch_screens(qtile):
    i = qtile.screens.index(qtile.current_screen)
    group = qtile.screens[i - 1].group
    qtile.current_screen.set_group(group)

dgroups_key_binder = None

dgroups_app_rules = []  # type: list

groups.append(ScratchPad("scratchpad", [
        DropDown("quake_term", terminal, height=0.4, width=.995, x=.0025, y=0, opacity=0.9, on_focus_lost_hide=False),
]))

mod = "mod4"

keys.extend([
    Key([mod], "j", lazy.layout.down(), desc="Move focus to next window"),
    Key([mod], "k", lazy.layout.up(), desc="Move focus to previous window"),
])

keys.extend([
    Key([mod, "shift"], "h", lazy.layout.shuffle_left(), desc="Move window to the left"),
    Key([mod, "shift"], "l", lazy.layout.shuffle_right(), desc="Move window to the right"),
    Key([mod, "shift"], "j", lazy.layout.shuffle_down(), desc="Move window down"),
    Key([mod, "shift"], "k", lazy.layout.shuffle_up(), desc="Move window up"),
])

keys.extend([
    Key([mod, "control"], "j", lazy.next_screen(), desc="Move focus to next monitor"),
    Key([mod, "control"], "k", lazy.prev_screen(), desc="Move focus to previous monitor"),
])

keys.append(Key([mod], "Tab", lazy.next_layout(), desc="Toggle between layouts"))

keys.extend([
    Key([mod], "h", lazy.layout.shrink_main(), desc="Shrink window to the left"),
    Key([mod], "l", lazy.layout.grow_main(), desc="Shrink window to the right"),
    Key([mod], "n", lazy.layout.normalize(), desc="Reset all window sizes"),
    Key([mod], "s", lazy.function(switch_screens), desc="Switch the groups on the two screens"),
])

keys.extend([
    Key([mod], "f", lazy.window.toggle_fullscreen(), desc="Toggle fullscreen on the focused window",),
    Key([mod], "t", lazy.window.toggle_floating(), desc="Toggle floating on the focused window"),
    Key([mod], "m", lazy.window.toggle_minimize(), desc="Toggle Minimize Window"),
    Key([mod, "shift"], "m", lazy.window.unminimize_all(), desc="Unminimize all windows in group"),
])

keys.extend([
    Key([mod, "shift"], "q", lazy.window.kill(), desc="Kill focused window"),
    Key([mod, "shift"], "r", lazy.reload_config(), desc="Reload the config"),
    Key([mod, "shift"], "e", lazy.shutdown(), desc="Shutdown Qtile"),
    Key([mod, "control"], "l", lazy.spawn(lockscreen), desc="Lock screen"),
    Key([mod, "control"], "h", lazy.spawn(hibernate), desc="Hibernate"),
    Key([mod, "control"], "s", lazy.spawn(suspend), desc="Suspend system"),
    Key([mod, "control"], "X", lazy.spawn("xkill"), desc="xkill"),
])

keys.extend([
    Key([mod], "r", lazy.spawncmd(), desc="Spawn a command using a prompt widget"),
    Key([mod], "d", lazy.spawn("rofi -dpi -show drun -theme ~/.config/rofi/rofi.rasi"),
        desc="Spawn a command using a prompt widget"),

    Key([mod], "Return", lazy.spawn(terminal), desc="Launch terminal"),
    Key([mod, "shift"], "Return", lazy.spawn(file_manager), desc="Launch File Manager"),
    Key([mod], "w", lazy.spawn(web_browser), desc="Launch Web Browser"),
    Key([mod], "e", lazy.spawn(emacs), desc="Launch Emacs"),
    Key([mod], "a", lazy.spawn("emacsclient --eval \"(emacs-everywhere)\""), desc="Use Emacs ANYWHERE"),
    Key([mod], "z", lazy.spawn("zotero"), desc="Launch Zotero"),

    Key(["control", "mod1"], "Delete", lazy.spawn(sysmon), desc="Launch System Monitor"),
    Key([mod], "Print", lazy.spawn("flameshot gui"), desc="Screenshot"),
    Key([mod], "p", lazy.spawn("flameshot gui"), desc="Screenshot"),
    Key([mod, "control"], "w", lazy.spawn(os.path.expanduser("~/.config/change-wallpaper.sh")), desc="Change Wallpaper"),
])

for i in groups:
    if len(i.name) == 1:
        keys.extend(
        [
                # mod1 + group number = switch to group
                Key(
                [mod],
                i.name,
                #lazy.group[i.name].toscreen(),      # the default behaviour
                lazy.function(go_to_group, i.name),  # my alternative function
                desc="Switch to group {}".format(i.name),
                ),
                # mod1 + shift + group number = switch to & move focused window to group
                Key(
                [mod, "shift"],
                i.name,
                lazy.window.togroup(i.name, switch_group=False),
                desc="Switch to & move focused window to group {}".format(i.name),
                ),
        ])

keys.extend([
        Key([mod, "control"], 'Return', lazy.group['scratchpad'].dropdown_toggle('quake_term')),
        ])

keys.extend([
    Key([], "XF86AudioRaiseVolume", lazy.spawn("amixer -D pulse sset Master 5%+"), desc="Raise Volume by 5%"),
    Key([], "XF86AudioLowerVolume", lazy.spawn("amixer -D pulse sset Master 5%-"), desc="Lower Volume by 5%"),
    Key([], "XF86AudioMute", lazy.spawn("amixer -D pulse set Master 1+ toggle"), desc="Toggle Mute Audio"),
    Key([], "XF86AudioNext", lazy.spawn("mpc next"), desc="Music Next"),
    Key([], "XF86AudioPrev", lazy.spawn("mpc prev"), desc="Music Previous"),
    Key([], "XF86AudioPlay", lazy.spawn("mpc toggle"), desc="Play/Pause Music"),
    Key([], "XF86AudioMicMute", lazy.spawn("amixer set Capture toggle"), desc="Toggle Mute Microphone"),
    Key([], "XF86MonBrightnessUp", lazy.spawn("light -A 10"), desc="Increase Brightness by 10%"),
    Key([], "XF86MonBrightnessDown", lazy.spawn("light -U 10"), desc="Decrease Brightness by 10%"),
])

mouse = [
    Drag([mod], "Button1", lazy.window.set_position_floating(), start=lazy.window.get_position()),
    Drag([mod], "Button3", lazy.window.set_size_floating(), start=lazy.window.get_size()),
    Click([mod], "Button2", lazy.window.bring_to_front()),
]

groups[8].matches = [Match(wm_class = "discord", title = "Discord Updater")]
groups[8].spawn = "discord"

groups[9].matches = [Match(wm_class = "spotify")]

floating_layout = layout.Floating(**floating_layout_theme,
    float_rules=[
        # Run the utility of `xprop` to see the wm class and name of an X client.
        *layout.Floating.default_float_rules,
        Match(wm_class="confirmreset"),  # gitk
        Match(wm_class="makebranch"),  # gitk
        Match(wm_class="maketag"),  # gitk
        Match(wm_class="ssh-askpass"),  # ssh-askpass
        Match(title="branchdialog"),  # gitk
        Match(title="pinentry"),  # GPG key password entry
        Match(title="emacs-everywhere"),
        Match(func=lambda c: ("REAPER" in c.info()['wm_class']) and (not "REAPER v" in c.info()['name'])),
    ],
)

@hook.subscribe.client_managed
def resize_floating(client):
    if "emacs-everywhere" in client.name:
        client.set_size_floating(1200,700)
        client.center()

if laptop:
        @hook.subscribe.client_managed
        def onboard_in_front(client):
                if client.name == "Onboard":
                        client.bring_to_front()

widget_defaults = dict(
    font="Ubuntu Nerd Font",
    fontsize=15,
    padding=3,
)
extension_defaults = widget_defaults.copy()

def init_widget_list(with_systray):
        widget_list = [
                        widget.GroupBox(
                                highlight_method = 'line',
                                disable_drag = True,
                                font = "K2D ExtraBold",
                                hide_unused = False,
                                highlight_color = [color_bar_background, color_bar_background], # background gradient
                                active = color_bar_foreground,
                                inactive = '505050', # font color
                                this_current_screen_border = color_primary,
                                this_screen_border = color_primary,
                                other_current_screen_border = None,
                                other_screen_border = None,
                                urgent_alert_method = 'line',
                                urgent_border = 'FF0000',
                                urgent_text = '000000',
                                use_mouse_wheel = False,
                                padding_x = 8 if laptop else None,
                                fontsize = 18 if laptop else 15,
                        ),
                        widget.Prompt(),
                        widget.Chord(
                                chords_colors={
                                        "launch": ("#ff0000", "#ffffff"),
                                },
                                name_transform=lambda name: name.upper(),
                        ),
                        widget.Spacer(),
                        widget.TaskList(
                                highlight_method = 'border',
                                border = color_primary,
                                borderwidth = 2,
                                unfocused_border = None,
                                max_title_width = 250,
                                markup_minimized = "<i>({})</i>",
                                markup_maximized = "<b>{}</b>",
                                txt_floating = "🗗 ",
                                txt_maximized = "🗖 ",
                                txt_minimized = "🗕 ",
                                margin_y = 6,
                                icon_size = 35,
                                stretch = False,
                                foreground = "FFFFFF",
                        ),
                        widget.Spacer(),
                        widget.WidgetBox(
                                close_button_location = 'right',
                                start_opened = False,
                                text_closed = '󰝡',
                                text_open = '󰝠',
                                fontsize = 20,
                                widgets=[widget.Systray(padding = 8,
                                        background = wal_background_versions[1] + bar_opacity)],
                                padding = 0,
                                foreground = color_bar_foreground,
                        ),
                        widget.Spacer(length=15),
                        widget.Clock(
                                format="  %H:%M, %a %-d. %b %y",
                                update_interval = 1.0,
                                padding = 0,
                                foreground = color_bar_foreground,
                        ),
                        widget.Spacer(length=16),
                        widget.Volume(
                                emoji = True,
                                emoji_list = ['', '', '', ' '],
                                foreground = color_bar_foreground,
                                padding = 0,
                        ),
                        widget.Volume(
                                emoji = False,
                                padding = 0,
                                foreground = color_bar_foreground,
                                mute_format = "",
                                unmute_format = '  {volume}%',
                        ),
                        widget.Spacer(length=13),
                        widget.Battery(
                                update_interval = 60,
                                charge_char = "󰂉",
                                discharge_char = "󰁿",
                                full_char = "󰁹",
                                empty_char = "󰂎",
                                not_charging_char = "󰂂",
                                unknown_char = "󰂑",
                                format = "{char} {percent:2.0%} / {hour:d}:{min:02d}h",
                                hide_threshold = None,
                                low_foreground = 'FF0000',
                                low_percentage = 0.11,
                                notify_below = 0.11,
                                notification_timeout = 0,
                                padding = 0,
                                foreground = color_bar_foreground,
                        ),
                        widget.Spacer(length=13),
                        widget.CurrentLayout(
                                foreground = color_bar_foreground,
                                padding = 0,
                                fontsize= 18,
                        ),
                        widget.Spacer(length=12),
                ]
        if not with_systray:
                # when counting list elements from the left, 
                # we want to pop them in descending order
                widget_list.pop(7) # spacer is eight widget from the left
                widget_list.pop(6) # systray is seventh widget from the left
        if config_in_use == "desktop":
                # when counting list elements from the right, 
                # we want to pop them in ascending order
                widget_list.pop(-4) # battery is fourth widget from the right
                widget_list.pop(-3) # spacer is third widget from the right
        return widget_list

my_bars = [bar.Bar(
            init_widget_list(with_systray),
            size = 40,
            background = color_bar_background, # this has bar_opacity baked in
            opacity = 1, # but we want no transparency of widgets
            border_width = 0,
            reserve = True,
            #margin = [5, 5, -2, 5],
        ) for with_systray in [True, False]]

screens = [
    Screen(top=my_bars[0]),
    Screen(top=my_bars[1]),
]

@hook.subscribe.startup_once
def autostart():
    home = os.path.expanduser('~/.config/qtile/autostart.sh')
    subprocess.Popen([home])
